tests/test_reweighting.py

```diff
@@ -0,0 +1,407 @@
+
+
+def make_gaussian_testsystem():
+    """normalized/unnormalized 1D Gaussian with a dependence on lambda and params"""
+
+    def annealed_gaussian_def(lam, params):
+        initial_mean, initial_log_sigma = 0.0, 0.0
+        target_mean, target_log_sigma = params
+
+        # lam = 0 -> (mean = 0, stddev = 1)
+        # lam = 1 -> (mean = target_mean, stddev = target_sigma)
+        mean = lam * target_mean - (1 - lam) * initial_mean
+        stddev = np.exp(lam * target_log_sigma + (1 - lam) * initial_log_sigma)
+
+        return mean, stddev
+
+    def sample(lam, params, n_samples):
+        mean, stddev = annealed_gaussian_def(lam, params)
+        return norm.rvs(loc=mean, scale=stddev, size=(n_samples, 1))
+
+    def logpdf(x, lam, params):
+        mean, stddev = annealed_gaussian_def(lam, params)
+        return np.sum(norm_logpdf(x, loc=mean, scale=stddev))
+
+    def u_fxn(x, lam, params):
+        """unnormalized version of -logpdf"""
+        mean, stddev = annealed_gaussian_def(lam, params)
+        return np.sum(0.5 * ((x - mean) / stddev) ** 2)
+
+    def normalized_u_fxn(x, lam, params):
+        return -logpdf(x, lam, params)
+
+    def reduced_free_energy(lam, params):
+        mean, stddev = annealed_gaussian_def(lam, params)
+        log_z = np.log(stddev * np.sqrt(2 * np.pi))
+        return -log_z
+
+    return u_fxn, normalized_u_fxn, sample, reduced_free_energy
+
+
+def assert_estimator_accurate(estimate_delta_f, analytical_delta_f, ref_params, n_random_trials, atol):
+    """for many random parameter sets, assert that the reweighted estimates of
+    delta_f(params) and grad(delta_f)(params) are consistent with analytical_delta_f function
+
+    Parameters
+    ----------
+    estimate_delta_f: Callable[[params], float]
+        estimated free energy difference as a
+        differentiable function of params
+    analytical_delta_f: Callable[[params], float]
+        exact free energy difference as a
+        differentiable function of params
+    ref_params: [2,] array
+        generate random parameters near ref_params
+    n_random_trials: int
+    atol: float
+        absolute tolerance
+    """
+
+    def sample_random_params():
+        mean = ref_params[0] + onp.random.rand()
+        log_sigma = ref_params[1] - onp.random.rand()
+        return np.array([mean, log_sigma])
+
+    f_hat, g_hat = value_and_grad(estimate_delta_f)(ref_params)
+    f_ref, g_ref = value_and_grad(analytical_delta_f)(ref_params)
+
+    onp.testing.assert_allclose(f_hat, f_ref, atol=atol)
+    onp.testing.assert_allclose(g_hat, g_ref, atol=atol)
+
+    for _ in range(n_random_trials):
+        trial_params = sample_random_params()
+        f_hat, g_hat = value_and_grad(estimate_delta_f)(trial_params)
+        f_ref, g_ref = value_and_grad(analytical_delta_f)(trial_params)
+
+        onp.testing.assert_allclose(f_hat, f_ref, atol=atol)
```

Comment:
atol makes sense here, since we're typically happy with a precision of say around 0.1kJ/mol